### ISSUE TITLE
[💡 Feature]: new `dialog` API namespace

### ISSUE THREAD

**Issue by christian-bromann** [webdriverio/webdriverio#11442]
### Is your feature request related to a problem?

Currently managing dialogs is not straightforward. While the WebDriver protocol provides commands like `acceptAlert` it still creates challenges, especially since it can cause automation to stale. Therefor I suggest to add a `dialog` namespace API that simplifies dealing with them.

### Describe the solution you'd like.

The current protocol commands `alertAccept`, `alertDismiss`, `sendAlertText` and `getAlertText` can be now called via:

```ts
browser.dialog.accept()
browser.dialog.dismiss()
browser.dialog.sendText()
browser.dialog.getText()
```

Next, I want to make use of WebdriverIOs preload script features which would allow us to mock these Web APIs which would allow users to not have to deal with them at all anymore.

```ts
browser.dialog.mock('alert' | 'confirm' | 'prompt' as T, opts: MockOptions[T])
browser.dialog.mock('alert', { autoDismiss?: boolean })
browser.dialog.mock('confirm', { confirmed: boolean })
browser.dialog.mock('prompt', { returnValue: string })
```

The above interface will immediately handle the dialog without even having it pop up. Users can mock all dialog calls or ones with certain messages, e.g.:

```ts
browser.dialog.mock('alert', {
  autoDismiss: true,
  message: 'Something happened!'
})
```

Above will only mock the following alert call: `window.alert('Something happened!')`.

Furthermore users can fetch alert calls and assert agains them:

```ts
console.log(await browser.dialog.calls())
```

could return:

```ts
[{
  type: 'alert',
  message: 'Something happened'
}, {
  type: 'confirm',
  message: 'Something happened',
  confirmed: true
}, {
  type: 'prompt',
  message: 'Something happened',
  defaultValue: 'foobar',
  returnValue: 'barfoo'
}]
```

All mocks that were initiated, can be disabled (for a certain message or all messages) via:

```ts
browser.dialog.mock('alert', {
  enabled: false,
  message: 'Something happened!' // optional
})
```

### Describe alternatives you've considered.

Keep status quo.

### Additional context

With WebDriver Bidi being available in Chrome and Firefox we can start adding some of these nice features. Of course in case this is not supported by a browser we will throw an error.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by erwinheitzman** [comment]
I love the idea and feel like this will help our users a lot, great suggestion!

What do you think about changing something like:
```js
browser.dialog.mock('alert', { autoDismiss?: boolean, message: 'Something happened' })
```
to something like
```js
browser.dialog.mock({ autoDismiss: true }) // dismiss all
browser.dialog('Something happened', { autoDismiss: true }) // dismiss spefic alert
```
I think this is cleaner as the message is not part of the mocked state or data.

I'm not sure if this is ever useful but do you think this would be a good addition to add?
```js
browser.dialog('Something happened', { message: 'Foo' }) // mocks the alert to show "Foo" instead of "Something happened"
```

**Comment by christian-bromann** [comment]
When it comes to:

```ts
browser.dialog.mock('alert', { autoDismiss: false, message: 'Something happened' }) // do not dismiss specific alert
// vs:
browser.dialog.mock('Something happened', { autoDismiss: false }) // do not dismiss specific alert
```

I would prefer first because it allows us to keep a consistent API which means the first parameter is always the type of the dialog, e.g. `alert`, `confirm` or `prompt`. Not all dialogs have a message so that part should stay optional.

> I'm not sure if this is ever useful but do you think the following example would be a good addition to add?

I am not sure this has a value for users.

**Comment by erwinheitzman** [comment]
The reason why I made my suggestion is because having the message in the object makes it seem like it's part of the mock while it is not so I find it confusing from a user perspective though I very much understand where you are coming from too.

Let's see what others think

**Comment by christian-bromann** [comment]
I see, that makes sense. Actually we would be more consistent by using parameters in `mock` to determine the mock and offer things like `mockReturnValue` to define how the behavior changes. Which means we would define the mock via:

```ts
const alert = browser.dialog.mock('alert') // for all alerts
const alert = browser.dialog.mock('alert', 'Something happened') // for alerts with the message 'Something happened'
const alert = browser.dialog.mock('alert', '') // for alerts with no message param (unlikely usage case)
```
and then have the user to call a function to define its behavior, e.g.:

```ts
alert.mockReturnValue(true) // enable auto dismiss
confirm.mockReturnValue(true)
prompt.mockReturnValue('some result')
```
This makes us consistent with other mock APIs. It doesn't for `alert` since there is no return value. I think it would make sense to say, when ever you define a mock, it will be dismissed automatically until you call: `alert.mockRestore()`.

@erwinheitzman what do you think?

**Comment by erwinheitzman** [comment]
I really like that @christian-bromann that looks great 👍

**Comment by nextlevelbeard** [comment]
Let's try and have an unifying API for mobile as well that deals with both iOS and Android alerts.

**Comment by christian-bromann** [comment]
@nextlevelbeard would that mean integrating https://github.com/webdriverio/appium-boilerplate/blob/0208c5583b4cd427ac4a87104649e53db60315c9/tests/screenobjects/components/NativeAlert.ts#L12 for iOS and Android sessions?

**Comment by nextlevelbeard** [comment]
> @nextlevelbeard would that mean integrating https://github.com/webdriverio/appium-boilerplate/blob/0208c5583b4cd427ac4a87104649e53db60315c9/tests/screenobjects/components/NativeAlert.ts#L12 for iOS and Android sessions?

That LGTM, I'm more concerned on how to confirm/dismiss alerts, iOS often has multiple ways to confirm/dismiss an alert: always or only once

**Comment by christian-bromann** [comment]
@nextlevelbeard can you elaborate on that please? What are your experiences on this?
